### PR TITLE
feat: make operator compatible with Ivy renderer

### PR DESCRIPTION
**Pull request by arturovt**
@NetanelBasal 

This PR represents compatibility with the Ivy renderer. Angular 9 is going to be out where the Ivy is enabled by default. Currently their API is exposed publicly with theta symbol.

Potentially this is a major feature and we could mention further in the readme that `ngx-take-until-destroy@6+` is compatible only with Angular 8+ and Ivy enabled.

### CODE REVIEW

**Comment by NetanelBasal** [comment]
Hi, thanks, I was planning to work on this soon. Unfortunately, the solution isn't working. We'll need to add a decorator and change it to something like:

```ts
export function UntilDestroy() {
  return cmpType => {
    let cmpInstance;
    const originalFactory = cmpType.ngComponentDef.factory;
    cmpType.ngComponentDef.onDestroy = () => {
      cmpInstance['__destroyed'].next();
      cmpInstance['__destroyed'].complete();
      if (cmpInstance.ngOnDestroy) {
        cmpInstance.ngOnDestroy();
      }
    };

    cmpType.ngComponentDef.factory = (...args) => {
      cmpInstance = originalFactory(...args);
      cmpInstance['__destroyed'] = new Subject();
      return cmpInstance;
    };
    return cmpType;
  };
}
```

**Comment by arturovt** [comment]
> Hi, thanks, I was planning to work on this soon. Unfortunately, the solution isn't working. We'll need to add a decorator and change it to something like:
> 
> ```ts
> export function UntilDestroy() {
>   return cmpType => {
>     let cmpInstance;
>     const originalFactory = cmpType.ngComponentDef.factory;
>     cmpType.ngComponentDef.onDestroy = () => {
>       cmpInstance['__destroyed'].next();
>       cmpInstance['__destroyed'].complete();
>       if (cmpInstance.ngOnDestroy) {
>         cmpInstance.ngOnDestroy();
>       }
>     };
> 
>     cmpType.ngComponentDef.factory = (...args) => {
>       cmpInstance = originalFactory(...args);
>       cmpInstance['__destroyed'] = new Subject();
>       return cmpInstance;
>     };
>     return cmpType;
>   };
> }
> ```

Interesting. But will the operator still be present?

**Comment by NetanelBasal** [comment]
Yes, it will be the same. Unfortunately, this is how Ivy works. Anyway, I thought about union this package and [ngx-auto-unsubscribe](https://github.com/NetanelBasal/ngx-auto-unsubscribe) to one, so that it will support both functionalities.

Do you want to do it?

**Comment by arturovt** [comment]
> Yes, it will be the same. Unfortunately, this is how Ivy works. Anyway, I thought about union this package and [ngx-auto-unsubscribe](https://github.com/NetanelBasal/ngx-auto-unsubscribe) to one, so that it will support both functionalities.
> 
> Do you want to do it?

As I understand you clearly we will have a package that will expose both `untilDestroyed` and `UntilDestroy` decorator that will work only together?

Sounds good. Much better than to mess around the already stable packages :smile:

**Comment by NetanelBasal** [comment]
Yes. It should be like:

```ts
@UntilDestroyed()
@Component({})
export class TestComponent {
  ngOnInit() {
    interval(1000)
      .pipe(untilDestroyed(this))
      .subscribe(console.log);
  }
}
```

Or they can choose to use the same functionality as `ngx-auto-unsubscribe`:

```ts
@UntilDestroyed()
@Component({})
export class TestComponent {
  ngOnInit() {
    this.interval = interval(1000)
  }
}
```

- It will be nice if you could create a new Angular CLI project and use `ng g library`
- Maybe create a migration script that adds the `@UntilDestroyed()` decorator.

**Comment by arturovt** [comment]
Okay, I will create a repo and work there.

Should we close this PR or keep it opened for the further discussion here?

**Comment by NetanelBasal** [comment]
Why create a new repo? You can work on this one.

**Comment by arturovt** [comment]
I've noticed that the `Subject` still should be created in the `untilDestroyed` operator. Because it's assigned to an instance after constructor invokation:
```ts
const instance = originalFactory();
instance['__destroy'] = new Subject...
...
```
The `__destroy` property is not available in the `constructor` yet.

So `UntilDestroy` decorator only has to override `def.onDestroy` property.

**Comment by NetanelBasal** [comment]
Yes, it could work either way. You can leave it in `untilDestroyed`.

**Comment by arturovt** [comment]
> You should run this block only if you have the following option:

```ts
UntilDestroy({
 checkProperties: true
})
```

Done

> upgrade to v6.0.0

Done

**Comment by arturovt** [comment]
> Why you didn't create the playground application? You created only the library.
> 
> `npx @angular-cli new until-destroy --enable-ivy=true`
> `ng g new library ngx-until-destroy`
> 
> That way we can add live examples to our playground

What do you think about adding E2E tests (using cypress)? This would help to make it clear that everything works as intended. I also could add an integration of the cypress + angular universal (ssr).

**Comment by NetanelBasal** [comment]
Sure, you can add it. But we don't need SSR.

**Comment by NetanelBasal** [comment]
Great work! What's our status?

File: integration/app/app.component.html [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<app-interval *ngIf="shown"></app-interval>
```
Comment (NetanelBasal) [inline integration/app/app.component.html]: We should try it with multiple instances and see that one doesn't affect the others.

**Comment by arturovt** [comment]
Updated README and provided multiple instances of components.

**Comment by arturovt** [comment]
@NetanelBasal

**Comment by NetanelBasal** [comment]
@arturovt are we ready?

**Comment by arturovt** [comment]
Yeah, we are.

**Comment by NetanelBasal** [comment]
Great, let me run some tests, update the README and the CI.

**Comment by NetanelBasal** [comment]
@arturovt did you test it with production build with AOT?

**Comment by NetanelBasal** [comment]
Maybe it worth to upgrade to [Angular 9](https://github.com/angular/angular/blob/master/CHANGELOG.md#900-next5-2019-09-04), and check it.

**Comment by arturovt** [comment]
> did you test it with production build with AOT?

Sure, I've added scripts to package.json to build production and serve those assets.

> Maybe it worth to upgrade to Angular 9, and check it.

Okay.

**Comment by NetanelBasal** [comment]
```json
  "devDependencies": {
    "@angular-builders/jest": "^8.2.0",
    "@angular-devkit/build-angular": "^0.803.2",
    "@angular/cli": "9.0.0-next.3",
    "@angular/common": "^9.0.0-next.5",
    "@angular/compiler": "^9.0.0-next.5",
    "@angular/compiler-cli": "^9.0.0-next.5",
    "@angular/core": "^9.0.0-next.5",
    "@angular/platform-browser": "^9.0.0-next.5",
    "@angular/platform-browser-dynamic": "^9.0.0-next.5",
    "@types/jest": "^24.0.18",
    "@types/node": "^12.7.3",
    "angular-http-server": "^1.9.0",
    "codelyzer": "^5.1.0",
    "jest": "^24.9.0",
    "ng-packagr": "^5.5.0",
    "rxjs": "^6.5.2",
    "tslint": "^5.19.0",
    "typescript": "3.5.3",
    "zone.js": "^0.10.2"
  }
```

It works, just update your code.

**Comment by arturovt** [comment]
Done

**Comment by arturovt** [comment]
Wait, they updated API in 9 verison

**Comment by arturovt** [comment]
Now there is no `def.factory` but static `ngFactoryDef` property.

**Comment by arturovt** [comment]
Ah, sorry, we don't override `def.factory`.

**Comment by arturovt** [comment]
@NetanelBasal what are we doing next?

**Comment by NetanelBasal** [comment]
I will take care of the README and fix the CI. Then, I will release it.

**Comment by NetanelBasal** [comment]
@arturovt, we missed a directive test. Can you please add it?

**Comment by arturovt** [comment]
> @arturovt, we missed a directive test. Can you please add it?

Added directive.

**Comment by NetanelBasal** [comment]
@arturovt, I want first to create the migration script. I'm working on it.

Can you add a validation that verifies the presence of the decorator on using `untilDestoryed`, please?

**Comment by NetanelBasal** [comment]
@arturovt I added the migration script. I test it on our project, and it works great. Can you test it on your project, please?

https://github.com/NetanelBasal/ngx-take-until-destroy/blob/master/migration/run.js

**Comment by arturovt** [comment]
> @arturovt I added the migration script. I test it on our project, and it works great. Can you test it on your project, please?
> 
> https://github.com/NetanelBasal/ngx-take-until-destroy/blob/master/migration/run.js

Sure, in the evening after job

**Comment by arturovt** [comment]
@NetanelBasal It's OK.

**Comment by NetanelBasal** [comment]
Great! Can you add a validation that verifies the presence of the decorator on using `untilDestoryed`, please?

**Comment by arturovt** [comment]
> Great! Can you add a validation that verifies the presence of the decorator on using `untilDestoryed`, please?

Hmm. We could add some property to the `ComponentDef` like `__untilDestroyDecorated = true` :D As we just override `onDestroy`. Are you ok with this approach or maybe you've got some other ideas? I'm open to any approach.

**Comment by NetanelBasal** [comment]
We can use Symbol() instead. Yes, I think it's worth the protection of memory leaks.

**Comment by NetanelBasal** [comment]
@arturovt I want to merge it to master without releasing it for now. Can you omit the README changes, please?

**Comment by arturovt** [comment]
@NetanelBasal hey. Reverted README and added a check if directive is decorated

**Comment by NetanelBasal** [comment]
@arturovt I released it. Can you please run a sanity check? To make sure everything works as expected when consuming the library?

**Comment by arturovt** [comment]
> @arturovt I released it. Can you please run a sanity check? To make sure everything works as expected when consuming the library?

Sure. Gimme some time

**Comment by Cole-Beckwith-SP** [comment]
Hey @NetanelBasal + @arturovt I'm curious why the Ivy changes require these changes. I've seen murmurs that Ivy collects the lifecycle hooks and does not actually call them, which makes the operators changes to ngOnDestroy ignored, but I haven't been able to find any official documentation as to why that happens.

**Comment by arturovt** [comment]
@NetanelBasal hey. Haven't you published the `@ngneat/until-destroy` to the npm yet? Cuz I'm getting 404 when installing `yarn add @ngneat/until-destroy`

**Comment by NetanelBasal** [comment]
Sorry, check now.

**Comment by arturovt** [comment]
@NetanelBasal I've tested and it works fine. Haven't found any issues. What's up from your side? I mean is everything OK and works as intented?

**Comment by NetanelBasal** [comment]
Yes, everything works fine. Thanks for your work!

**Comment by arturovt** [comment]
@NetanelBasal 

Hey.

Have you seen this PR https://github.com/angular/angular/pull/33110.

As I understand we'll need to change `ngDirectiveDef` in the future as it gets merged and released?

**Comment by NetanelBasal** [comment]
It seems like. We need to track it.

**Pull request merged**